llama.cpp

```diff
@@ -947,8 +969,30 @@ struct llama_hparams {
     float rope_freq_base_train;
     float rope_freq_scale_train;
 
+    bool operator==(const llama_hparams & other) const {
+        if (this->vocab_only != other.vocab_only) return false;
+        if (this->n_vocab != other.n_vocab) return false;
+        if (this->n_ctx_train != other.n_ctx_train) return false;
+        if (this->n_embd != other.n_embd) return false;
+        if (this->n_head != other.n_head) return false;
+        if (this->n_head_kv != other.n_head_kv) return false;
+        if (this->n_layer != other.n_layer) return false;
+        if (this->n_rot != other.n_rot) return false;
+        if (this->n_ff != other.n_ff) return false;
+
+        const float EPSILON = 1e-9;
+
+        if (!is_float_eq(this->f_norm_eps, other.f_norm_eps, EPSILON)) return false;
+        if (!is_float_eq(this->f_norm_rms_eps, other.f_norm_rms_eps, EPSILON)) return false;
+        if (!is_float_eq(this->rope_freq_base_train, other.rope_freq_base_train, EPSILON)) return false;
+        if (!is_float_eq(this->rope_freq_scale_train, other.rope_freq_scale_train, EPSILON)) return false;
+
+        return true;
+    }
+
+    // implement != explicitly using the "==" implementation above so we don't get a warning about it
     bool operator!=(const llama_hparams & other) const {
-        return static_cast<bool>(memcmp(this, &other, sizeof(llama_hparams))); // NOLINT
+        return !(*this == other);
```

Comment:
Oops, I forgot that C++ doesn't automatically infer operators based on the others like Python does. Simpler to just keep `!=` then, since that's the only operator we seem to need.